File: pieces.rb
class Piece
  attr_accessor :board, :color, :name, :move_coords, :pos
  def initialize(options)
    @board = options[:board]
    @color = options[:color]
    @name = options[:name]
    @pos = options[:pos]
    @move_coords = []
  end

  def valid_moves
    moves.select do |move|
      !move_into_check?(move)
    end
  end

  def move_into_check?(pos)
    new_board = Board.new(self.board.dup)
    new_board.move(self.pos,pos, true)
    return true if new_board.checked?(self.color)
  end
end

class InvalidMoveException < StandardError
end

class SlidingPiece < Piece
  def move_dirs(cord)

    moves = []
    i =  1
    while true
      pos = [self.pos[0] + (cord[0]* i), self.pos[1] + (cord[1] * i)]
      break if self.board.off_the_grid?(pos)
      hit_piece = self.board.pieces[pos[0]][pos[1]]
      if !hit_piece.nil?
        moves << pos if hit_piece.color != self.color
        break
      end
      moves << pos
      i += 1
    end

    moves
  end

  def moves
    self.move_coords.inject([]) do |moves,dir|
      moves + move_dirs(dir)
    end
  end
end

class SteppingPiece < Piece
  def moves
    self.move_coords.map do |move|
      [pos[0] + move[0],pos[1] + move[1]]
    end.select do |move|
      next if self.board.off_the_grid?(move)
      space = self.board.pieces[move[0]][move[1]]
      space.nil? || space.color != self.color
    end
  end
end

class Pawn < SteppingPiece
  attr_accessor :passant
  def initialize(options)
    super(options)
    @passant = false
  end
  def to_s
    self.color == :white ? "\u2659" : "\u265F"
  end
  def moves
    moves = []

    [[1,1],[1,-1]].each do |attack|
      attack[0] = -attack[0] if self.color == :white
      space = self.board.pieces[pos[0] + attack[0]][pos[1] + attack[1]]
      moves << [pos[0] + attack[0], pos[1] + attack[1]] unless space.nil? || space.color == self.color
    end
    last = self.board.last_moved
    if last.is_a?(Pawn) && last.passant && (last.pos[1]-self.pos[1]).abs == 1
      moves << [last.pos[0]+1,last.pos[1]] if self.color == :black && self.pos[0] == 4
      moves << [last.pos[0]-1,last.pos[1]] if self.color == :white && self.pos[0] == 3
    end

    steps = [[1,0]]
    steps << [2,0] if self.color == :white && self.pos[0] == 6
    steps << [2,0] if self.color == :black && self.pos[0] == 1

    steps.each do |move|
      move[0] = -move[0] if self.color == :white
      space = self.board.pieces[pos[0] + move[0]][pos[1] + move[1]]
      moves << [pos[0] + move[0],pos[1] + move[1]] if space.nil?
    end

    moves
  end
end

class Knight < SteppingPiece
  KNIGHT_MOVES = [[1,2],[-1,2],[1,-2],
    [-1,-2],[2,1],[-2,1],[-2,-1],[2,-1]]

  def initialize(options)
    super
    @move_coords = KNIGHT_MOVES
  end

  def to_s
    self.color == :white ? "\u2658" : "\u265E"
  end
end

class King < SteppingPiece
  attr_accessor :moved
  KING_MOVES = [[1, 0], [0, 1], [-1, 0], [0, -1],
    [1, 1], [-1, 1], [-1, -1], [1, -1], [0, 2], [0, -2]]

  def initialize(options)
    super
    @move_coords = KING_MOVES
    @moved = false
  end

  def to_s
    self.color == :white ? "\u2654" : "\u265A"
  end
end

class Bishop < SlidingPiece
  BISHOP_MOVES = [[1, 1], [-1, 1], [-1, -1], [1, -1]]

  def initialize(options)
    super
    @move_coords = BISHOP_MOVES
  end

  def to_s
    self.color == :white ? "\u2657" : "\u265D"
  end
end

class Rook < SlidingPiece
  attr_accessor :moved
  ROOKIE_MOVES = [[1, 0], [0, 1], [-1, 0], [0, -1]]

  def initialize(options)
    super
    @moved = false
    @move_coords = ROOKIE_MOVES
  end

  def to_s
    self.color == :white ? "\u2656" : "\u265C"
  end
end

class Queen < SlidingPiece
  QUEEN_MOVES = [[1, 0], [0, 1], [-1, 0], [0, -1],
    [1, 1], [-1, 1], [-1, -1], [1, -1]]

  def initialize(options)
    super
    @move_coords = QUEEN_MOVES
  end

  def to_s
    self.color == :white ? "\u2655" : "\u265B"
  end
end

File: web/js/pieces.js
var Piece = Class.extend({
  init: function(options) {
    this.board = options.board;
    this.color = options.color;
    this.pos = options.pos;
  },
  valid_moves: function(){
    var moves = this.moves();
    var valid_moves = [];
    for (var i = 0; i < moves.length; i++) {
      if (!this.move_into_check(moves[i]))
        valid_moves.push(moves[i]);
    }
    return valid_moves;
  },
  move_into_check : function(pos) {
    var new_board = new Board(this.board.dup());//new Board( J
    new_board.move(this.pos, pos, true);
    return new_board.checked(this.color);
  }
});

var SlidingPiece = Piece.extend({
  init: function(options) {
    this._super(options);
  },
  move_dirs: function(cord) {
    var moves = [];
    var i = 1;

    while (true) {
      var pos = [this.pos[0] + (cord[0] * i), this.pos[1] + (cord[1] * i)];
      if (this.board.off_the_grid(pos)) {
        break;
      }
      var hit_piece = this.board.pieces[pos[0]][pos[1]];
      if (hit_piece) {
        if (hit_piece.color != this.color) {
          moves.push(pos);
          break;
        }
      }
      moves.push(pos);
      i++;
    }
    return moves;
  },
  moves: function() {
    var moves = [];
    var coords = this.move_coords;
    for (var i= 0; i < coords.length; i++) {

      moves = moves.concat(this.move_dirs(coords[i]));
    }

    return moves;
  }
});

var SteppingPiece = Piece.extend({
  init: function(options) {
    this._super(options);
  },
  moves: function() {
    var move_arr = [];
    for (var i in this.move_coords) {
      var x = this.move_coords[i][0] + this.pos[0];
      var y = this.move_coords[i][1] + this.pos[1];
      move_arr.push([x, y]);
    }
    var ret_arr = [];
    for (var i in move_arr) {
      if (!this.board.off_the_grid(move_arr[i])) {
        var space = this.board.pieces[move_arr[i][0]][move_arr[i][1]];
        if (!space || space.color != this.color) {
        	ret_arr.concat(move_arr[i]);
        }
      }
    }
    return move_arr;
  }
});

var Pawn = SteppingPiece.extend({
  init: function(options) {
    this._super(options);
    this.passant = false;
  },
  to_s: function() {
    return this.color == "white" ? "\u2659" : "\u265F";
  },
  moves: function() {
    var moves = [];
    var attack_arr = [[1,1], [1,-1]];
    for (var i in attack_arr) {
      if (this.color == "white") {
        attack_arr[i][0] = -attack_arr[i][0];
      }
      var space = this.board.pieces[this.pos[0] + attack_arr[i][0]][this.pos[1] + attack_arr[i][1]];
      if (space && space.color != this.color) {
        moves.push([this.pos[0] + attack_arr[i][0], this.pos[1] + attack_arr[i][1]]);
      }
    }
    var last = this.board.last_moved;
    if (last instanceof Pawn && last.passant && Math.abs((last.pos[1]-this.pos[1])) == 1) {
      if (this.color == "black" && this.pos[0] == 4) {
        moves.push([last.pos[0]+1,last.pos[1]]);
      } else if (this.color == "white" && this.pos[0] == 3) {
        moves.push([last.pos[0]-1,last.pos[1]]);
      }
    }
    var steps = [[1, 0]];
    if ((this.color == "white" && this.pos[0] == 6) || (this.color == "black" && this.pos[1] == 1)) {
      steps.push([2,0]);
    }
    for (var i in steps) {
      if (this.color == "white") {
        steps[i][0] = -(steps[i][0]);
      }
      var space = this.board.pieces[this.pos[0] + steps[i][0]][this.pos[1] + steps[i][1]];
      if (!space) {
        moves.push([this.pos[0] + steps[i][0],this.pos[1] + steps[i][1]]);
      }
    }
    return moves;
   }
});

var Knight = SteppingPiece.extend({
    init: function(options) {
      this._super(options);
      this.move_coords = [[1,2],[-1,2],[1,-2],
      [-1,-2],[2,1],[-2,1],[-2,-1],[2,-1]];
    },

    to_s: function() {
      return this.color == "white" ? "\u2658" : "\u265E";
    }
});

var King = SteppingPiece.extend({
  init: function(options) {
    this._super(options);
    this.king = true;
    this.moved = false;
    this.move_coords = [[1, 0], [0, 1], [-1, 0], [0, -1],
    [1, 1], [-1, 1], [-1, -1], [1, -1], [0, 2], [0, -2]];
  },
  to_s: function() {
    return this.color == "white" ? "\u2654" : "\u265A";
  }
});

var Bishop = SlidingPiece.extend({
  init: function(options) {
    this._super(options);
    this.move_coords = [[1, 1], [-1, 1], [-1, -1], [1, -1]];
  },
  to_s: function() {
    return this.color == "white" ? "\u2657" : "\u265D";
  }
});

var Rook = SlidingPiece.extend({
  init: function(options) {
    this._super(options);
    this.moved = false;
    this.move_coords = [[1, 0], [0, 1], [-1, 0], [0, -1]];
  },
  to_s: function() {
    return this.color == "white" ? "\u2656" : "\u265C";
  }
});

var Queen = SlidingPiece.extend({
  init: function(options) {
    this._super(options);
    this.move_coords = [[1, 0], [0, 1], [-1, 0], [0, -1],
    [1, 1], [-1, 1], [-1, -1], [1, -1]];
  },
  to_s: function() {
    return this.color == "white" ? "\u2655" : "\u265B";
  }
});